### ISSUE TITLE
getSynchronizationSources (and getContriburingSources) should work for video too

### ISSUE THREAD

**Issue by henbos** [w3c/webrtc-pc#1983]
Multiple things in the spec hint that this is only for audio RTP streams, e.g.

- "When the first audio frame contained in an RTP packet is delivered to the RTCRtpReceiver's MediaStreamTrack for playout, the user agent MUST queue a task to update the relevant information..."
- audioLevel says that audioLevel must be calculated based on the audio data even if the header extension containing this value is missing, concluding that the member must never be absent for SSRCs. This would not be a valid conclusion if getSynchronizationSources() was applicable to video RTP streams too.

The spec should be updated to support both audio and video cases. The "source" (SSRC or CSRC) and timestamp are both valuable pieces of information for a receiver, whether receiving audio or video. For example, they can be used to figure out whose ("source" used as an identifier) stream is currently being received or if a stream is "active" (if timestamp is recent). These APIs are designed to work in real-time. Trying to get this information through other means such as getStats() is not performance-viable.

Any reason this was written for only audio in mind?

**Comment by henbos** [comment]
@fippo  was wondering why you don't just look at MediaStreamTrack.muted to figure out if the stream is active or not but while unmuting happens as a result of an RTP packet arriving, muting only happens as a result of negotiation (setRemoteDescription's muteTracks list).

If you want timestamps for RTP packets to determine how active your stream is or what its "source" is there already is an appropriate API, so let's enable this for video too. It already takes care of the the fact that we can only update this info per task execution cycle or else we expose the fact that it is not single-threaded.

**Comment by fippo** [comment]
> but while unmuting happens as a result of an RTP packet arriving, muting only happens as a result of negotiation (setRemoteDescription's muteTracks list).
I do not think that is true, see https://w3c.github.io/webrtc-pc/#mediastreamtrack-network-use
And I am not happy with muted flipflopping for both network and api events.

getSynchronizationSources() and getContributingSources() was built for audio-levels mainly (which is also the reason it does not make much sense for video -- i do not think any video rtp mixer uses csrcs...). If we want a timestamp for last-packet-received we should not use this API but expose something on the receiver.

**Comment by fippo** [comment]
what getContributingSources was designed for is this (jitsi, probably from around 2011):
![csrc](https://user-images.githubusercontent.com/289731/45491114-2bdaaa80-b769-11e8-8c49-48b60e0437ca.jpg)
(from wikipedia via @emcho)
-- showing the audio levels for participants, even when audio is mixed and the mixer uses the csrc audio level. getSynchronizationSources does the same but is a better alternative to using webaudio or even getStats to get this. Overloading this with the new semantics that you want for video does not seem like a good idea.

(and I did not want the "timestamp last packet received" on the receiver, not the track)

**Comment by jan-ivar** [comment]
@henbos [timestamp](http://w3c.github.io/webrtc-pc/#dom-rtcrtpcontributingsource-timestamp) is poorly named here: It indicates *"the most recent time of* **playout** *of media"*.

It's predicted playout time through a media element. It's ***not*** the RTP timestamp.

As @fippo [mentions](https://github.com/w3c/webrtc-pc/issues/1983#issuecomment-421010310): this API was designed specifically for audio bars per participant. *getStats* was deemed too slow for this purpose, which is why these synchronous methods exists.

Are there compelling use-cases for runtime playout time estimation of video?
